src/storage/patch/SparqlUpdatePatchHandler.ts

```diff
@@ -41,11 +41,43 @@ export class SparqlUpdatePatchHandler extends PatchHandler {
     // Verify the patch
     const { identifier, patch } = input;
     const op = patch.algebra;
-    if (!this.isDeleteInsert(op)) {
+    this.validateUpdate(op);
+
+    const lock = await this.locker.acquire(identifier);
+    try {
+      await this.applyPatch(identifier, op);
+    } finally {
+      await lock.release();
+    }
+  }
+
+  private isDeleteInsert(op: Algebra.Operation): op is Algebra.DeleteInsert {
+    return op.type === Algebra.types.DELETE_INSERT;
+  }
+
+  private isComposite(op: Algebra.Operation): op is Algebra.CompositeUpdate {
+    return op.type === Algebra.types.COMPOSITE_UPDATE;
+  }
+
+  /**
+   * Checks if the input operation is of a supported type (DELETE/INSERT or composite of those)
+   */
+  private validateUpdate(op: Algebra.Operation): void {
+    if (this.isDeleteInsert(op)) {
+      this.validateDeleteInsert(op);
+    } else if (this.isComposite(op)) {
+      this.validateComposite(op);
+    } else {
```

Comment:
Minor implementation suggestion: we're now at the tipping point where, if it gets any more complex, we likely want `Operation` classes that can self-validate and self-apply.

So rather than having the `if` statements here and below, we could to `operation.validate` and `operation.apply`, and they know what to do for themselves. Because now we basically have inline polymorphism 🙂

I'll leave it to you whether you think it's already worth it.

Reply:
Yea I was also thinking if one more operation is added this should be generalized. But for 2 it seemed fine (since we're probably not adding any more soon). The sparql algebra library has some helper functions that could help with this.

Reply:
Actually had an initial look into this since extracting the permissions has a similar issue, but a decent implementation of this is probably going to get quite some work since there are a lot of nested objects here. So either we create a new highly nested object with the same structure as the algebra object, or we continuously create new objects on the fly. Or we don't use objects but have some utility functions to iterate through the objects.